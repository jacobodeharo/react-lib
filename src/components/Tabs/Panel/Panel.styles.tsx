import styled from 'styled-components';

const StyledPanel = styled.div`
  background-color: var(--c-panel-bg, ${({ theme }) => theme.primaryColorLight});
  color: var(--c-panel-fg, ${({ theme }) => theme.primaryColorText});
  padding: 10px;
  border-radius: 0 0 5px 5px;
`;

export { StyledPanel };
