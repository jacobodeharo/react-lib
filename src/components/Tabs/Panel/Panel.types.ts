import { Theme } from '@/components/theme';

export type PanelProps = {
  tabId: string;
  theme?: Theme;
};
