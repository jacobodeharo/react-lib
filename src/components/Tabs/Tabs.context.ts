import { TTabsContext } from './Tabs.types';
import * as React from 'react';

const TabsContext = React.createContext<TTabsContext | undefined>(undefined);

export { TabsContext };
