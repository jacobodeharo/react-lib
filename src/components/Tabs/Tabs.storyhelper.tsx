import { Anchor, Button } from '../..';
import React from 'react';

const renderPanelContentA = () => (
  <article>
    <h4>Tab A</h4>
    <p>
      Nemo enim ipsam voluptatem quia voluptas sit aspernatur aut odit aut
      fugit, sed quia consequuntur magni dolores eos qui ratione voluptatem
      sequi nesciunt. Neque porro quisquam est, qui dolorem ipsum quia dolor sit
      amet, consectetur, adipisci velit, sed quia non numquam eius modi tempora
      incidunt ut labore et dolore magnam aliquam quaerat voluptatem.
    </p>
  </article>
);

const renderPanelContentB = () => (
  <article>
    <h4>Tab B</h4>
    <p>
      Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium
      doloremque laudantium, totam rem aperiam, eaque ipsa quae ab illo
      inventore veritatis et quasi architecto beatae vitae dicta sunt explicabo.
      <br />
      Nemo enim ipsam voluptatem quia voluptas sit aspernatur aut odit aut
      fugit, sed quia consequuntur magni dolores eos qui ratione voluptatem
      sequi nesciunt. Neque porro quisquam est, qui dolorem ipsum quia dolor sit
      amet, consectetur, adipisci velit, sed quia non numquam eius modi tempora
      incidunt ut labore et dolore magnam aliquam quaerat voluptatem.
    </p>
    <Button onClick={() => alert('It works!')}>Push me!</Button>
  </article>
);

const renderPanelContentC = () => (
  <article>
    <h4>Tab C</h4>
    <p>
      Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent id
      tellus augue. Proin ut enim maximus, elementum enim sed, semper augue.
      Nulla facilisi. Suspendisse potenti. Donec efficitur orci vitae efficitur
      faucibus. Etiam a mi et sapien pretium venenatis. Proin ac tellus nec
      metus fringilla elementum eget quis est. Ut finibus sed tortor id
      interdum. Mauris ullamcorper lorem ac dapibus imperdiet. Ut libero turpis,
      varius sit amet risus non, dignissim venenatis tortor. Curabitur non
      sollicitudin est. Donec pretium quam est, dignissim blandit felis porta
      blandit. In et lorem eu lectus finibus fringilla quis at purus. Sed porta
      ac velit sit amet finibus.
    </p>
    <Anchor href='http://google.com' rel='noopener noreferer' target='_blank'>
      Press this link to search in Google
    </Anchor>
  </article>
);

export { renderPanelContentA, renderPanelContentB, renderPanelContentC };
