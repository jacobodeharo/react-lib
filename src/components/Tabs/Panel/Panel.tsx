import React from 'react';
import { PanelProps } from './Panel.types';
import { useTabs } from '../Tabs.hooks';

const Panel: React.FC<PanelProps> = (props) => {
  const { activeTab } = useTabs();
  const { children, tabId } = props;
  return activeTab === tabId ? <div>{children}</div> : null;
};

export { Panel };
