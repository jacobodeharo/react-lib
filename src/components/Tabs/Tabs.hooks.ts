import React from 'react';
import { TTabsContext } from './Tabs.types';
import { TabsContext } from './Tabs.context';

const useTabs = (): TTabsContext => {
  const context = React.useContext(TabsContext);
  if (!context) {
    throw new Error('This component must be used within a <Tabs> component.');
  }
  return context;
};

export { useTabs };
