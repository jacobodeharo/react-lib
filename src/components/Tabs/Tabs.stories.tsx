import React from 'react';
import { Meta, Story } from '@storybook/react/types-6-0';
import { ThemeProvider } from '../theme';
import { availableThemes, ThemeID } from '../theme/variants';
import { Tabs } from './Tabs';
import { TabsProps } from './Tabs.types';

import { Selector } from './Selector/Selector';
import { Panel } from './Panel/Panel';
import {
  RenderPanelContentA,
  RenderPanelContentB,
  RenderPanelContentC,
} from './Tabs.storyhelper';

export default {
  title: '@/components/Containers/Tabs',
  component: Tabs,
  argTypes: {
    themeName: {
      control: {
        type: 'select',
        options: Object.keys(availableThemes),
      },
    },
  },
} as Meta;

type StoryTabsProps = TabsProps & {
  themeName: string;
};

const Template: Story<StoryTabsProps> = (args) => {
  const themeTextcolor =
    availableThemes[args.themeName as ThemeID].primaryTextColor;
  return (
    <ThemeProvider theme={availableThemes[args.themeName as ThemeID]}>
      <Tabs selectedTab='Tab A'>
        <Selector tabId='Tab A'>Tab A</Selector>
        <Selector tabId='Tab B'>Tab B</Selector>
        <Selector tabId='Tab C'>Tab C</Selector>
        <Panel tabId='Tab A'>{RenderPanelContentA(themeTextcolor)}</Panel>
        <Panel tabId='Tab B'>{RenderPanelContentB(themeTextcolor)}</Panel>
        <Panel tabId='Tab C'>{RenderPanelContentC(themeTextcolor)}</Panel>
      </Tabs>
    </ThemeProvider>
  );
};

export const Primary = Template.bind({});
Primary.args = {
  themeName: 'gray',
};
