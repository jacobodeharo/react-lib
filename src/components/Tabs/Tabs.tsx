import * as React from 'react';
import { TabsProps } from './Tabs.types';
import { TabsContext } from './Tabs.context';
import { Selector } from './Selector/Selector';
import { Panel } from './Panel/Panel';
import { renderChildrenOfType } from '../../helper/reactDomUtils';
import { TabsBody, TabsHeader, TabsWrapper } from './Tabs.styles';

const Tabs: React.FC<TabsProps> = (props) => {
  const { children, selectedTab } = props;
  const [activeTab, setActiveTab] = React.useState(selectedTab || '');
  const memoizedContextValue = React.useMemo(
    () => ({
      activeTab,
      setActiveTab,
    }),
    [activeTab, setActiveTab]
  );
  return (
    <TabsContext.Provider value={memoizedContextValue}>
      <TabsWrapper>
        <TabsHeader>{renderChildrenOfType(children, Selector)}</TabsHeader>
        <TabsBody>{renderChildrenOfType(children, Panel)}</TabsBody>
      </TabsWrapper>
    </TabsContext.Provider>
  );
};

export { Tabs };
