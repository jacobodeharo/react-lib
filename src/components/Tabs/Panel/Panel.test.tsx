import React from 'react';
import { render, cleanup, screen } from '@testing-library/react';
import { Panel } from './Panel';
import { AMBER_THEME } from '../../theme/variants/amberTheme';
import { consoleLogSpy } from '../../../helper/testHelper';
import { ErrorBoundary } from '../../../core/ErrorBoundary/ErrorBoundary';
import { TabsContext } from '../Tabs.context';

describe('Panel', () => {
  const PANEL_CHILDREN = 'Panel Label';
  afterEach(cleanup);

  test('has to be rendered', () => {
    render(
      <TabsContext.Provider
        value={{
          activeTab: 'tabA',
          setActiveTab: (_: string) => undefined,
        }}
      >
        <Panel tabId='tabA'>{PANEL_CHILDREN}</Panel>
      </TabsContext.Provider>
    );
    const PanelElement = screen.getByText(PANEL_CHILDREN);
    expect(PanelElement).toBeInTheDocument();
    expect(PanelElement).toMatchSnapshot();
  });

  test('has to throw an error if it is not rendered in a Tabs component', () => {
    const spy = consoleLogSpy();
    render(
      <ErrorBoundary>
        <Panel tabId='tabA'>{PANEL_CHILDREN}</Panel>
      </ErrorBoundary>
    );
    expect(spy).toHaveBeenCalled();
    spy.mockRestore();
  });

  test('has to accept theming', () => {
    render(
      <TabsContext.Provider
        value={{
          activeTab: 'tabA',
          setActiveTab: (_: string) => undefined,
        }}
      >
        <Panel tabId='tabA' theme={AMBER_THEME}>
          {PANEL_CHILDREN}
        </Panel>
      </TabsContext.Provider>
    );
    const PanelElement = screen.getByText(PANEL_CHILDREN);
    expect(PanelElement).toBeInTheDocument();
    expect(PanelElement).toMatchSnapshot();
  });
});
