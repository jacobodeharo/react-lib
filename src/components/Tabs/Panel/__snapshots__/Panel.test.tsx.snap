// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Panel has to accept theming 1`] = `
<div
  class="sc-bdnylx cFAyRC"
>
  Panel Label
</div>
`;

exports[`Panel has to be rendered 1`] = `
<div
  class="sc-bdnylx cFAyRC"
>
  Panel Label
</div>
`;
